src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            //
+            // class C1
+            // {
+            //   public C2 C2 { get; set; } = new C2();
+            //   public void M()
+            //   {
+            //     var x = new C1 { C2 = { P1 = 1, P2 = 2 } };
+            //   }
+            // }
+            // class C2
+            // {
+            //   public int P1 { get; set; }
+            //   public int P2 { get; set; }
+            // }
+            //
+            // We generate the following code for C1.M(). Note the multiple calls to C1::get_C2().
+            //   IL_0000: nop
+            //   IL_0001: newobj instance void C1::.ctor()
+            //   IL_0006: dup
+            //   IL_0007: callvirt instance class C2 C1::get_C2()
+            //   IL_000c: ldc.i4.1
+            //   IL_000d: callvirt instance void C2::set_P1(int32)
+            //   IL_0012: nop
+            //   IL_0013: dup
+            //   IL_0014: callvirt instance class C2 C1::get_C2()
+            //   IL_0019: ldc.i4.2
+            //   IL_001a: callvirt instance void C2::set_P2(int32)
+            //   IL_001f: nop
+            //   IL_0020: stloc.0
+            //   IL_0021: ret
+            //
+            // We therefore visit the InitializedMember to get the implicit receiver for the contained initializer, and that implicit receiver will be cloned everywhere it encounters
+            // an IInstanceReferenceOperation with ReferenceKind InstanceReferenceKind.ImplicitReceiver
+
+            IOperation implicitReceiver = Visit(operation.InitializedMember);
+            HandleObjectOrCollectionInitializer(operation.Initializer, implicitReceiver);
+            return null;
+        }
+
+        private void HandleObjectOrCollectionInitializer(IObjectOrCollectionInitializerOperation initializer, IOperation implicitReceiver)
+        {
+            // PROTOTYPE(dataflow): Handle collection initializers
+            IOperation previousImplicitReceiver = _currentImplicitReceiver;
+            _currentImplicitReceiver = implicitReceiver;
+
+            foreach (IOperation innerInitializer in initializer.Initializers)
+            {
+                AddStatement(Visit(innerInitializer));
+            }
+
+            _currentImplicitReceiver = previousImplicitReceiver;
+        }
+
+        public override IOperation VisitInstanceReference(IInstanceReferenceOperation operation, int? captureIdForResult)
+        {
+            if (operation.ReferenceKind == InstanceReferenceKind.ImplicitReceiver)
+            {
+                // When we're in an object or collection initializer, we need to replace the instance reference with a reference to the object being initialized
+                Debug.Assert(_currentImplicitReceiver != null);
+                Debug.Assert(operation.IsImplicit);
+                return OperationCloner.CloneOperation(_currentImplicitReceiver);
```

Comment:
>_currentImplicitReceiver [](start = 54, length = 24)

It looks like this could be an arbitrary expression that might include nodes that are only created by this builder. Currently, OperationCloner throws on many of them. I think we need to implement cloning for them after all.